pkg/mimir/signals.go

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package mimir
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/grafana/dskit/services"
+	"github.com/weaveworks/common/signals"
+	"go.uber.org/atomic"
+)
+
+// newShutdownSignalReceiver creates a new signals.SignalReceiver implementation that
+// handles everything required to cleanly shut down Mimir in response to a SIGTERM or
+// SIGINT signal.
+func newShutdownSignalReceiver(delay time.Duration, ready *atomic.Bool, server *http.Server, manager *services.Manager) signals.SignalReceiver {
+	return &shutdownSignalReceiver{
+		delay:   delay,
+		ready:   ready,
+		server:  server,
+		manager: manager,
+	}
+}
+
+// shutdownSignalReceiver takes care of the process of cleanly shutting down Mimir
+// in the event of a SIGTERM or SIGINT signal.
+//
+// * Flips an atomic boolean (which should be used elsewhere to force readiness checks to fail).
+// * Disables HTTP keep-alives so that pooled connections to this instance are closed.
+// * Optionally, wait for a configured delay before actually stopping any services.
+// * Stops all running Mimir services via a services.Manager.
+type shutdownSignalReceiver struct {
+	delay   time.Duration
+	ready   *atomic.Bool
+	server  *http.Server
+	manager *services.Manager
+}
+
+func (r *shutdownSignalReceiver) Stop() error {
+	r.ready.Store(false)
+	r.server.SetKeepAlivesEnabled(false)
+
+	if r.delay > 0 {
+		time.Sleep(r.delay)
+	}
+
+	r.manager.StopAsync()
```

Comment:
I'd suggest to keep this logic in https://github.com/grafana/mimir/blob/55fd8d949ef0a97af9216d6eaf348d7462f081a2/pkg/mimir/mimir.go#L805-L807 goroutine to keep it together and simple. This type and file is too shallow to exist imho.

Reply:
OK. I can move the call to `manager.StopAsync()` back and move the type to `mimir.go`

Reply:
I'm not sure we even need extra type. Can we move the code to the goroutine function directly, or perhaps to a method on `*Mimir`?

Reply:
We could but it felt more correct to me to use the `SignalReceiver` abstraction from the weaveworks signal handler since this is what it was intended for.

Reply:
It may be more "correct", but it splits the signal handling and shutdown logic (which is already complicated enough) to multiple places. Personally I would keep it together in `Run` method as it was before, without adding more indirection via signal receiver.

Reply:
Done in a8a84a44e9c062dfba700b443e6f1485c82c5b85. You are right, this feels easier to follow.

Reply:
Thank you!